topdown/walk.go

```diff
@@ -50,39 +76,36 @@ func walk(filter, path *ast.Array, input *ast.Term, iter func(*ast.Term) error)
 	switch v := input.Value.(type) {
 	case *ast.Array:
 		for i := 0; i < v.Len(); i++ {
-			path = pathAppend(path, ast.InternedIntNumberTerm(i))
-			if err := walk(filter, path, v.Elem(i), iter); err != nil {
+			if err := walk(filter, pathAppend(path, ast.InternedIntNumberTerm(i)), v.Elem(i), iter); err != nil {
 				return err
 			}
-			path = path.Slice(0, path.Len()-1)
 		}
 	case ast.Object:
 		return v.Iter(func(k, v *ast.Term) error {
-			path = pathAppend(path, k)
-			if err := walk(filter, path, v, iter); err != nil {
+			if err := walk(filter, pathAppend(path, k), v, iter); err != nil {
 				return err
 			}
-			path = path.Slice(0, path.Len()-1)
 			return nil
 		})
 	case ast.Set:
-		return v.Iter(func(elem *ast.Term) error {
-			path = pathAppend(path, elem)
-			if err := walk(filter, path, elem, iter); err != nil {
+		for _, elem := range v.Slice() {
+			if err := walk(filter, pathAppend(path, elem), elem, iter); err != nil {
 				return err
 			}
-			path = path.Slice(0, path.Len()-1)
-			return nil
-		})
+		}
 	}
 
 	return nil
 }
 
-var emptyArr = ast.ArrayTerm()
-
 func walkNoPath(input *ast.Term, iter func(*ast.Term) error) error {
-	if err := iter(ast.ArrayTerm(emptyArr, input)); err != nil {
+	arr := arrayPool.Get().(*ast.Array)
+	defer arrayPool.Put(arr)
```

Comment:
We could wrap that into a function that logs to verify when it's actually run:
```go
defer func() {
    log.Println("put an array back")
     arrayPool.Put(arr)
}()
```
as per my previous comment, I'm not sure this ever runs before the end of the evaluation....of either the entire query or the expression, I'm not entirely sure about that.

Reply:
Yeah, I'm also not certain so I'll play around some more with this. I was surprised it worked at all here, but it actually does, whereas the regular walk function doesn't like it at all, and would have both the path and the values come back with some surprising results.

Reply:
See my comment above..